File: Baekjoon/Class3/24.2805 나무 자르기.py
from collections import Counter
import sys
input = sys.stdin.readline

def is_length_enough(cutter_height, required, trees):
    length = 0
    for tree, count in trees:
        if tree > cutter_height:
            length += (tree - cutter_height) * count
            if length >= required: #충분한 길이
                return True
    return False #충분하지 않음

def max_cutter_height(required, trees):
    answer = 0
    left = 0
    right = max(tree for tree, _ in trees)
    while left <= right:
        middle = (left + right) // 2
        if is_length_enough(middle, required, trees): #더 높게 자를 수도 있음
            left = middle + 1
            answer = middle
        else: #더 낮게 잘라야 함
            right = middle - 1
    return answer

def main():
    _, m = map(int, input().split())
    trees = list(Counter(map(int, input().split())).items())
    print(max_cutter_height(m, trees))

main()

File: Baekjoon/Class3/24.2805 나무 자르기.js
const inputFile = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(inputFile, "utf8").trim().split("\n");

const isLengthEnough = (cutterHeight, required, trees) => {
    let length = 0
    for(const tree of trees)
        if(tree > cutterHeight) {
            length += tree - cutterHeight;
            if(length >= required) return true; // 충분한 길이
        }
    return false; // 충분하지 않음
}

const maxCutterHeight = (required, trees) => {
    let answer = 0;
    let left = 0, right = Math.max(...trees);
    while(left <= right) {
        const middle = Math.trunc((left + right) / 2);
        if(isLengthEnough(middle, required, trees)) { // 더 높게 자를 수도 있음
            left = middle + 1;
            answer = middle;
        }
        else // 더 낮게 잘라야 함
            right = middle - 1;
    }
    return answer;
}

const [_, m] = input[0].split(" ").map(Number);
const trees = input[1].split(" ").map(Number);
console.log(maxCutterHeight(m, trees));